plugins/MobileMessaging/API.php

```diff
@@ -80,57 +79,101 @@ public function setSMSAPICredential($provider, $credentials = array())
         $settings[MobileMessaging::API_KEY_OPTION] = $credentials;
 
         $this->model->setCredentialManagerSettings($settings);
-
-        return true;
     }
 
     /**
-     * add phone number
+     * Adds a phone number for the current user
      *
      * @param string $phoneNumber
+     * @return void
+     */
+    public function addPhoneNumber(string $phoneNumber): void
+    {
+        Piwik::checkUserIsNotAnonymous();
+
+        $phoneNumber = $this->sanitizePhoneNumber($phoneNumber);
+
+        $phoneNumbers = $this->model->getPhoneNumbers(Piwik::getCurrentUserLogin(), false);
+
+        $unverifiedPhoneNumbers = array_filter(
+            $phoneNumbers,
+            function ($phoneNumber) {
+                return !$phoneNumber['verified'];
+            }
+        );
+
+        if (count($unverifiedPhoneNumbers) >= 3) {
+            throw new \Exception('You cannot add more than 3 unverified phone numbers!');
+        }
+
+        $this->sendVerificationCodeAndAddPhoneNumber($phoneNumber);
+    }
+
+    /**
+     * Requests a new verification code for the given phone number
      *
-     * @return bool true
+     * @param string $phoneNumber
+     * @return void
      */
-    public function addPhoneNumber($phoneNumber)
+    public function resendVerificationCode(string $phoneNumber): void
     {
         Piwik::checkUserIsNotAnonymous();
 
         $phoneNumber = $this->sanitizePhoneNumber($phoneNumber);
 
-        $verificationCode = "";
-        for ($i = 0; $i < self::VERIFICATION_CODE_LENGTH; $i++) {
-            $verificationCode .= Common::getRandomInt(0, 9);
+        $phoneNumbers = $this->model->getPhoneNumbers(Piwik::getCurrentUserLogin(), false);
+
+        if (empty($phoneNumbers[$phoneNumber])) {
+            throw new \Exception("The phone number $phoneNumber has not yet been added.");
+        }
+
+        if (true === $phoneNumbers[$phoneNumber]['verified']) {
+            throw new \Exception("The phone number $phoneNumber has already been verified.");
         }
 
+        if ($phoneNumbers[$phoneNumber]['requestTime'] > time() - 60) {
+            throw new \Exception("A verification code for phone number $phoneNumber has already been requested recently.");
+        }
+
+        $this->sendVerificationCodeAndAddPhoneNumber($phoneNumber);
+    }
+
+    private function sendVerificationCodeAndAddPhoneNumber(string $phoneNumber): void
+    {
+        $verificationCode = Common::getRandomString(6, 'abcdefghijklmnoprstuvwxyz0123456789');
```

Comment:
Hm... that was actually just a copy of the default value of `Common::getRandomString`. Interesting that we didn't use a `q` in random strings since 2010 for whatever reason 🙈